oeps/oep-0007.rst

```diff
@@ -0,0 +1,526 @@
+All packages should maintain the proper trove classifiers for the versions of
+Python they support.
+
+In the following recommendations, the major version classifiers comprise:
+
+.. code::
+
+    Programming Language :: Python :: 2
+    Programming Language :: Python :: 2 :: Only
+    Programming Language :: Python :: 3
+    Programming Language :: Python :: 3 :: Only
+
+Minor version classifiers include, but are not limited to:
+
+.. code::
+
+    Programming Language :: Python :: 2.6
+    Programming Language :: Python :: 2.7
+    Programming Language :: Python :: 3.5
+    Programming Language :: Python :: 3.6
+
+Packages that do not yet support Python 3 should list both of the major version
+Python 2 classifiers, plus any minor version classifiers that apply.
+
+Packages that support both Python 2 and Python 3 should include major version
+classifiers for both versions of Python, but must not include either of the
+``:: Only`` classifiers.
+
+Packages that have dropped Python 2 support should list both of the major
+version Python 3 classifiers, plus any minor version classifiers that apply.
+
+Ideally, all listed minor versions should be tested in a continuous integration
+environment.  At a minimum, at least the lowest and highest minor versions of
+each supported major version must be tested.
+
+Other problems
+--------------
+
+If you find other incompatibilities, a shim will likely be found as part of
+``six``.  For incompatibilies with no other solution, edX will maintain a
+repository of compatibility shims (edx-compat?).  Ideally, all edx-maintained
+code that implements different behavior based on Python version will be in this
+repo.
+
+When writing code that explicitly switches based on version, do
+
+.. code:: python3
+
+    if six.PY2:  # or future.PY2
+        do_python2_thing()
+    else:
+        do_python3_thing()
+
+Do not explicitly call out ``six.PY3`` or ``future.PY3``.  This should be more
+future-compatible with a potential future Python 4.
+
+
+Changing Coding Practice
+++++++++++++++++++++++++
+
+Changing internal code practices to ease conversion will require a
+three-pronged approach of documentation, socialization, and tooling.  To start,
+we need to update the official edX code style guide to mandate compatible code
+practices.  To socialize these practices among our engineers, we will announce
+our efforts to migrate to Python 3 during an engineering all-hands meeting,
+offer a workshop in writing compatible code, and promote awareness of
+incompatibilities during code reviews.  Additionally, we will host regular
+Python 3 office hours to help answer questions and troubleshoot problems that
+arise during migration.
+
+Appropriate tooling will help.  Tests should be configured to run under both
+Python 2 or 3 using tox.  As it would double the cost of running
+tests to run both all the time, we may only want to enable Python 3 tests in
+the CI environment for repos that are actively being migrated.  We still want
```

Comment:
Please don't worry about this. We are building our CI infrastructure for scale. If we reach a point where we can't support what we need to do, then we can think about removing pieces...